### ISSUE TITLE
Support for case sensitivity settings

### ISSUE THREAD

**Issue by piotr-oles** [streamich/memfs#533]
Hi!
As you probably know, there are 2 options for case sensitivity: 
 * Windows use case sensitive paths, 
 * other systems like macOS or Linux use case insensitive paths

I checked how `memfs` behave and it seems that it uses case sensitive approach:
```javascript
const { fs } = require('memfs');

fs.writeFileSync('/my/Case/sensitive/Path.txt', 'content');

console.log(fs.readFileSync('/my/case/sensitive/path.txt'));
// logs undefined
// if I would use native fs on macOS or Linux, I would get "content" logged
// so it's kind of inconsistency between memfs and fs
```
It would be nice if:
 1. there would be an option to configure it
 2. the default behavior would be to use system settings

**Comment by streamich** [comment]
> 1. there would be an option to configure it

Yes, it seems we could add support for case insensitive paths.

> 2. the default behavior would be to use system settings

I'm not sure if `memfs` should "magically" change the way it handles paths; I would prefer it is set explicitly. But happy to hear opinions.

**Comment by G-Rath** [comment]
I might be missing something, but:

> As you probably know, there are 2 options for case sensitivity:

The OSs act the opposite way around from your list: Windows is case insensitive, while UNIX is case sensitive.

![image](https://user-images.githubusercontent.com/3151613/81512725-bf1d1680-9376-11ea-88bd-cbd1dbb2455c.png)

> I'm not sure if memfs should "magically" change the way it handles paths

For now I agree with that - I think `memfs` should be careful adopting "magical" features since they tend to be very "gain this, lose that", but an explicit option could work.

I think in this case actually the OS side should be ignored, since the reason why `fs` is/isnt case-sensitive is because of that; aka it's not Node going "oh you're on Linux", it's instead a result of the underlying API calls V8 makes to the native file system.

Because of that, I think you could make a case for saying something like having the current behavior being the "default", and then implementing a `ignoreCase` option that causes memfs to just call `toLowerCase` on everything?

**Comment by G-Rath** [comment]
On an aside: I think `memfs` should probably favor defaulting to unix behaviours since it already uses the unix file path format (i.e `/my/path/` instead of `C:\my\path`)

**Comment by piotr-oles** [comment]
> I might be missing something, but:
>
> > As you probably know, there are 2 options for case sensitivity:
>
> The OSs act the opposite way around from your list: Windows is case insensitive, while UNIX is case sensitive.

Yes, I was wrong about Windows, but still, macOS is case-insensitive 😃 

> Because of that, I think you could make a case for saying something like having the current behavior being the "default", and then implementing a ignoreCase option that causes memfs to just call toLowerCase on everything?

Yes, Implementing it just as a configuration option would be enough 👍

**Comment by G-Rath** [comment]
> macOS is case-insensitive 

Interesting, that is very useful to know given that most of our devs at work are on macs while our servers are Ubuntu 😄 😬

**Comment by piotr-oles** [comment]
Is it something that I could expect to be implemented by one of the maintainers, or you would like to get a PR from me? 😃

**Comment by streamich** [comment]
@piotr-oles PR from you :)

**Comment by ajafff** [comment]
Turns out this is not that hard to implement. I opened #632 to make case sensitivity configurable.
